app/interactors/schedule/save_datetime_interactor.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+class Schedule::SaveDatetimeInteractor
+  include Interactor
+  delegate :params,
+           :edition,
+           :user,
+           :issues,
+           :datetime,
+           to: :context
+
+  def call
+    Edition.transaction do
+      find_and_lock_edition
+      check_for_issues
+      set_scheduled_publishing_datetime
+    end
+  end
+
+private
+
+  def find_and_lock_edition
+    context.edition = Edition.lock.find_current(document: params[:document])
+  end
+
+  def check_for_issues
+    unless edition.draft? || edition.submitted_for_review?
+      raise "Can't save a scheduling date unless edition is a draft or has been submitted for 2i"
+    end
+
+    checker = Requirements::ScheduledDatetimeChecker.new(schedule_params)
+    pre_issues = checker.pre_submit_issues.issues
+
+    issues = Requirements::CheckerIssues.new(pre_issues + action_issues)
+
+    context.fail!(issues: issues) if issues.any?
+
+    context.datetime = checker.parsed_datetime
+  end
+
+  def schedule_params
+    params.require(:schedule).permit(:day, :month, :year, :time)
+  end
+
+  def set_scheduled_publishing_datetime
+    updater = Versioning::RevisionUpdater.new(edition.revision, user)
+    updater.assign(scheduled_publishing_datetime: context.datetime)
+
+    context.fail! unless updater.changed?
+
+    edition.assign_revision(updater.next_revision, user).save!
+  end
+
+  def action_issues
+    return [] if scheduled_datetime_already_exists?
```

Comment:
So `already_scheduled?` wouldn't be quite right because the user doesn't need to have scheduled the document for publishing in order to change the scheduled_publishing_datetime.  They could have just set it already and then gone in and changed it again and not scheduled it.  According to the prototype if a scheduled_publishing_datetime has been set and then they go and change it again there is no choice of action to take; they just save the changed scheduled_publishing_datetime.

We'll be covering changing the scheduled_publishing_datetime after a document has been scheduled in this card https://trello.com/c/5n7oZvXZ/851-change-a-scheduled-publishing-date

Reply:
Ah, that's a shame - I was hopeful it would be simple. I think this will come up again if we try to address the 'Back' button bug (the assumption will break under that condition), so I'm happy to leave it - I'm afraid it's still confusing for me to read, both here and in the view, so it'd be great to have a deeper think about what the concept behind this check is.